icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1144,11 +1141,41 @@ public static String getName(String localeID){
         } else if ("root".equalsIgnoreCase(localeID)) {
             tmpLocaleID = EMPTY_STRING;
         } else {
-            tmpLocaleID = UND_PATTERN.matcher(localeID).replaceFirst(EMPTY_STRING);
+            tmpLocaleID = stripLeadingUnd(localeID);
         }
         return nameCache.getInstance(tmpLocaleID, null /* unused */);
     }
 
+    /**
+     * Strips out the leading "und" language code case-insensitively.
+     *
+     * @implNote Avoids creating new local non-primitive objects to reduce GC pressure.
+     */
+    private static String stripLeadingUnd(String localeID) {
+        int length = localeID.length();
+        if (length < 3) {
+            return localeID;
+        }
+
+        // If not starts with "und", return.
+        if (!localeID.regionMatches(true /* ignoreCase */, 0, "und", 0, 3 /* len */)) {
+            return localeID;
+        }
+
+        // The string is equals to "und" case-insensitively
+        if (length == 3) {
+            return EMPTY_STRING;
+        }
+
+        // localeID must have a length >= 4
+        char forthChar = localeID.charAt(3);
```

Comment:
Spelling: fourth

But a better name might be something like _separator_, to describe the meaning instead of the position of this character.